compiler-hacking.md

```diff
@@ -0,0 +1,464 @@
+# Understanding the Shemsu compiler
+This is a guide to understanding the Shesmu olive compiler. This is meant to
+explain the core components that get the language to Java bytecode.
+
+Since the compiler will continue to expand and change, the bytecode dumps in
+this tutorial may be different from future output of the compiler. However, the
+core behaviour of the Shesmu compiler is based on Java's `Stream`, which is
+stable, so they will be functionally similar. Bytecode dumps can be seen on the
+olive dashboard.
+
+To start off, compiling a simple script:
+
+     Input shesmu;
+
+is compiled to:
+
+    public class dyn/shesmu/Program extends ca/on/oicr/gsi/shesmu/ActionGenerator {
+    
+      public <init>()V
+        ALOAD 0
+        INVOKESPECIAL ca/on/oicr/gsi/shesmu/ActionGenerator.<init> ()V
+        RETURN
+        MAXSTACK = 0
+        MAXLOCALS = 1
+    
+      private clearGauge()V
+        RETURN
+        MAXSTACK = 0
+        MAXLOCALS = 1
+    
+      public synchronized run(Lca/on/oicr/gsi/shesmu/ActionConsumer;Ljava/util/function/Function;)V
+        ALOAD 0
+        INVOKEVIRTUAL dyn/shesmu/Program.clearGauge ()V
+       L0
+       L1
+        RETURN
+        MAXSTACK = 0
+        MAXLOCALS = 3
+    
+      public static <clinit>()V
+        RETURN
+        MAXSTACK = 0
+        MAXLOCALS = 0
+    }
+
+This class, `dyn.shesmu.Program`, extends the `ActionGenerator` superclass,
+which is used by the server to execute actions. The class name is reused for
+every script, but since each is loaded by a separate class loader, they are not
+in conflict. The constructor is relatives boring; it just calls the
+super-constructor. The `run` method is where all the interesting things happen.
+The `clearGauge` method is executed at the start and performs all the necessary
+initialisation for gauges. This is a convenience of compiler design: the `run`
+and `clearGauge` methods can be built in parallel even though they are executed
+sequentially. Since this script has no olives, the `run` method does nothing
+else and returns.
+
+Let's move to a script with an olive:
+
+    Input shesmu;
+    Olive
+      Run nothing With value = type;
+
+
+The output bytecode is more complicated. The initialisation is much the same:
+    
+    public class dyn/shesmu/Program extends ca/on/oicr/gsi/shesmu/ActionGenerator {
+    
+      public <init>()V
+        ALOAD 0
+        INVOKESPECIAL ca/on/oicr/gsi/shesmu/ActionGenerator.<init> ()V
+        RETURN
+        MAXSTACK = 0
+        MAXLOCALS = 1
+    
+      private clearGauge()V
+        RETURN
+        MAXSTACK = 0
+        MAXLOCALS = 1
+    
+      public synchronized run(Lca/on/oicr/gsi/shesmu/ActionConsumer;Ljava/util/function/Function;)V
+        ALOAD 0
+        INVOKEVIRTUAL dyn/shesmu/Program.clearGauge ()V
+
+Now there are interesting differences. First, the current time is stored for timing the olive runtime.
+
+       L0
```

Comment:
It wasn't immediately clear to me that this was an extension of the previous code sample, could the entire thing be provided beforehand or the commentary moved to code comments?